Firewall_rules diff mode support

omnom62 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
This change adds support for [diff mode](https://docs.ansible.com/ansible/latest/user_guide/playbooks_checkmode.html#using-diff-mode) to vyos_firewall_rules. This can aide someone into understanding what will check when using [check mode](https://docs.ansible.com/ansible/latest/user_guide/playbooks_checkmode.html#using-check-mode), and can help someone know what has changed when running their play (using just --diff without --check).

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->
This change deprecates https://github.com/vyos/vyos.vyos/pull/222

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
vyos_firewall_rules

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
Running a playbook in with DIFF mode on yields a report on changes (if any):
`$  ansible-playbook -i hosts -l vyos1.5 firewall_rules.yaml --diff --check

PLAY [Testing all the modules] **************************************************************************************************************************************************************************

TASK [Overrides all device configuration with provided configuration] ***********************************************************************************************************************************
[firewall ipv4 name Downlink rule]
+ 502 {
+     action "reject"
+     description "Rule 502 is configured by Ansible"
+     ipsec {
+         match-ipsec-out
+     }
+ }


[edit]
changed: [vyos1.5]

PLAY RECAP **********************************************************************************************************************************************************************************************
vyos1.5                    : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0`
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Test results
<!--
Provide the output of the unit tests and confirmation of the sanity tests
along with a description of which versions of VyOS you have tested against.

Tests will be run before the PR is accepted, but do not run automatically
on forks, so please run all of the tests with each modification of your PR
to ensure they will pass.

Unit test information can be found in the [Ansible Unit Tests](https://docs.ansible.com/ansible/latest/dev_guide/testing_units.html#testing-units)
section of the documentation.

Sanity test information can be found in the [Ansible Sanity Tests](https://docs.ansible.com/ansible/latest/dev_guide/testing_sanity.html#testing-sanity)
section of the Ansible documentation.

```
Example:

$ ansible-tests units
============================= test session starts ==============================
platform linux -- Python 3.12.2, pytest-8.1.1, pluggy-1.4.0
rootdir: /root/ansible_collections/vyos/vyos
configfile: ../../../ansible/test/lib/ansible_test/_data/pytest/config/default.ini
plugins: xdist-3.5.0, mock-3.14.0
created: 24/24 workers
24 workers [244 items]

........................................................................ [ 29%]
........................................................................ [ 59%]
........................................................................ [ 88%]
............................                                             [100%]
- generated xml file: /root/ansible_collections/vyos/vyos/tests/output/junit/python3.12-controller-units.xml -
============================= 244 passed in 1.55s ==============================

Describe the versions of VyOS that you have tested your changes
against.

```
-->
- [x] Sanity tests passed
- [x] Unit tests passed

Tested against VyOS versions:
<!-- examples, add or delete as appropriate; if using rolling versions, please specify
    fully
-->
- 1.3.8
- 1.4-rolling-202201010100
- 1.5-2025.03.18-0018-rolling


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [ ] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the ansible sanity and unit tests
- [ ] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added unit tests to cover my changes
- [x] I have added a file to `changelogs/fragments` to describe the changes

gaige commented:
Do we have any unit or integration tests for the diff mode?

omnom62 commented:
> Do we have any unit or integration tests for the diff mode?

I did not put together any - What could they be?  I was trying to design one, but could only come up with a string comparison for the output...

gaige commented:
> > Do we have any unit or integration tests for the diff mode?
> 
> I did not put together any - What could they be? I was trying to design one, but could only come up with a string comparison for the output...

I would think we'd want to have tests for:
- Diff with no change
- Diff with change
- Diff with multiple changes in the same rule
- Diff with changes in multiple rules.

It's going to basically be a string check, since this is an output generating function. I think we have the output we need to do it in unit tests, although might make sense to add to integration test as well (at least one variant).

omnom62 commented:
> > > Do we have any unit or integration tests for the diff mode?
> > 
> > 
> > I did not put together any - What could they be? I was trying to design one, but could only come up with a string comparison for the output...
> 
> I would think we'd want to have tests for:
> 
> * Diff with no change
> * Diff with change
> * Diff with multiple changes in the same rule
> * Diff with changes in multiple rules.
> 
> It's going to basically be a string check, since this is an output generating function. I think we have the output we need to do it in unit tests, although might make sense to add to integration test as well (at least one variant).

No sure we can do this in unit tests given the diff flag cannot be specified inside the test case code but only at the execution level (specifically, to test the real funtionality).
I think I can try and test this feature in integration tests instead

omnom62 commented on tests/integration/targets/vyos_firewall_rules/vars/pre-v1_4.yaml:
```diff
@@ -128,3 +128,126 @@ state_dict:
+              source:
+                address: 192.0.2.0
+              state: "{{ state_dict }}"
+  diff: '[edit firewall name INBOUND rule 101]\n>action reject\n\n[edit]'
```
Thanks @gaige - tried a few variants >- |- and the like - Ansible interpreter tends to add extra \\ when evaluating the string with \n and other escape chars 
I keep testing, I may need to represent the output string as a list or dict to compare later

gaige replied:
Hmm, I tried just now using the current source and this seemed to work for replaced_diff_02 (pre-1.4 test):

```yaml
  diff: |-
    [edit firewall name INBOUND rule 102]
    >action accept
    >protocol udp
    
    [edit]
````

omnom62 replied:
All done

gaige commented on changelogs/fragments/firewall_rules_diff.yml:
```diff
@@ -0,0 +1,3 @@
+---
+trivial:
```
This should likely be `minor_changes`

omnom62 replied:
done

gaige commented on tests/integration/targets/vyos_firewall_rules/vars/v1_4.yaml:
```diff
@@ -120,3 +120,138 @@ deleted_afi_all:
+              source:
+                address: 192.0.2.0
+              state: "{{ state_dict }}"
+  diff: >
```
These would be much easier to read using `|-`, do we have a reason to keep the explicit line breaks with escapes?

omnom62 replied:
thanks - I oversight this option, now fixed

gaige reviewed (approved):
Looks good. Thanks!

dmbaturin reviewed (approved):
Logic seems correct to me and I trust Gaige's judgment.

RubenNL commented:
This merge request seems to break integration tests on VyOS 1.3.4. This is the only version in the 1.3 range I have, so I can't verify others.

This is the error:

```
TASK [vyos_firewall_rules : Assert - Diff for a single rule and attribute] *****
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
replaced_diff_01.diff.rstrip() }}"
fatal: [test]: FAILED! => {
    "assertion": "result.diff.prepared == \"{{ replaced_diff_01.diff.rstrip() }}\"",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

I debugged a bit more:
`result.diff.prepared`: `[edit firewall name INBOUND rule 101]\n>action reject\n\n\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000[edit]`
`replaced_diff_01.diff.rstrip()`: `[edit firewall name INBOUND rule 101]\n>action reject\n\n[edit]`

So, something seems to be adding a lot of `\u0000`. (zero bytes)

omnom62 commented:
> This merge request seems to break integration tests on VyOS 1.3.4. This is the only version in the 1.3 range I have, so I can't verify others. 1.5 doesn't have any errors.
> 
> This is the error:
> 
> ```
> TASK [vyos_firewall_rules : Assert - Diff for a single rule and attribute] *****
> [WARNING]: conditional statements should not include jinja2 templating
> delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
> replaced_diff_01.diff.rstrip() }}"
> fatal: [test]: FAILED! => {
>     "assertion": "result.diff.prepared == \"{{ replaced_diff_01.diff.rstrip() }}\"",
>     "changed": false,
>     "evaluated_to": false,
>     "msg": "Assertion failed"
> }
> ```
> 
> I debugged a bit more: `result.diff.prepared`: `[edit firewall name INBOUND rule 101]\n>action reject\n\n\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000[edit]` `replaced_diff_01.diff.rstrip()`: `[edit firewall name INBOUND rule 101]\n>action reject\n\n[edit]`
> 
> So, something seems to be adding a lot of `\u0000`. (zero bytes)

Thanks @RubenNL 
I just run against 1.3.4 (even though it should not be used for testing. 1.3.8 was said to be the oldest valid one), and here are the results:
`$ ansible-test network-integration --inventory ~/lab/ansible/hosts vyos_firewall_rules 
Running vyos_firewall_rules integration test role
<skipped>
TASK [vyos_firewall_rules : include common vars] *******************************
ok: [target01]

TASK [vyos_firewall_rules : Setup 1.3.4] ***************************************
changed: [target01]

TASK [vyos_firewall_rules : Replace device configurations - No Diff] ***********
ok: [target01]

TASK [vyos_firewall_rules : Assert No Diff] ************************************
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [vyos_firewall_rules : Replace single rule's attribute and register Diff] ***
[edit firewall name INBOUND rule 101]
>action reject

[edit]
changed: [target01]

TASK [vyos_firewall_rules : Assert - Diff for a single rule and attribute] *****
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
replaced_diff_01.diff.rstrip() }}"
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [vyos_firewall_rules : Replace single rule's multiple attributes and register Diff] ***
[edit firewall name INBOUND rule 102]
>action accept
>protocol udp

[edit]
changed: [target01]

TASK [vyos_firewall_rules : Assert - Diff for a single rule and multiple attributes] ***
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
replaced_diff_02.diff.rstrip() }}"
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [vyos_firewall_rules : Replace attributes in multiple rules and register Diff] ***
[edit firewall ipv6-name UPLINK rule 1]
>action reject
>protocol udp
[edit firewall name INBOUND rule 102]
>protocol tcp

[edit]
changed: [target01]

TASK [vyos_firewall_rules : Assert - Diff for a single rule and multiple attributes] ***
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
replaced_diff_03.diff.rstrip() }}"
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}
<skipped>
TASK [vyos_firewall_rules : include common vars] *******************************
ok: [target01]

TASK [vyos_firewall_rules : Setup 1.3.4] ***************************************
changed: [target01]

TASK [vyos_firewall_rules : Replace device configurations - No Diff] ***********
ok: [target01]

TASK [vyos_firewall_rules : Assert No Diff] ************************************
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [vyos_firewall_rules : Replace single rule's attribute and register Diff] ***
[edit firewall name INBOUND rule 101]
>action reject

[edit]
changed: [target01]

TASK [vyos_firewall_rules : Assert - Diff for a single rule and attribute] *****
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
replaced_diff_01.diff.rstrip() }}"
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [vyos_firewall_rules : Replace single rule's multiple attributes and register Diff] ***
[edit firewall name INBOUND rule 102]
>action accept
>protocol udp

[edit]
changed: [target01]

TASK [vyos_firewall_rules : Assert - Diff for a single rule and multiple attributes] ***
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
replaced_diff_02.diff.rstrip() }}"
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [vyos_firewall_rules : Replace attributes in multiple rules and register Diff] ***
[edit firewall ipv6-name UPLINK rule 1]
>action reject
>protocol udp
[edit firewall name INBOUND rule 102]
>protocol tcp

[edit]
changed: [target01]

TASK [vyos_firewall_rules : Assert - Diff for a single rule and multiple attributes] ***
[WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: result.diff.prepared == "{{
replaced_diff_03.diff.rstrip() }}"
ok: [target01] => {
    "changed": false,
    "msg": "All assertions passed"
}
`
Looks OK to me.
What found weird in your output is, these symbols "\u0000" - It may suggest some encoding issues with Ansible / YAML interpreter

RubenNL commented:
You are correct about the versions, but I'm still waiting on a license I should already have been granted :( This is the best I an do.

I am able to fix the issue by installing `ansible-pylibssh` using `pip3 install ansible-pylibssh`. After removing it, I again get the error.

The fallback when `pylibssh` isn't available is `paramiko`, which I've been using since I started using this module.

Interestingly, `ansible-pylibssh` was removed from the `requirements.txt` of this module in this commit: https://github.com/vyos/vyos.vyos/pull/352/commits/54e0d62b65deb4993f73571b7ea0ca9138f79e03

My guess is that your environment is configured to use `ansible-pylibssh`.

To verify, please force Ansible to use paramiko: `ANSIBLE_NETWORK_CLI_SSH_TYPE=paramiko ansible-test network-integration --inventory ~/lab/ansible/hosts vyos_firewall_rules` I expect this to fail, just like it did for me.

omnom62 commented:
> You are correct about the versions, but I'm still waiting on a license I should already have been granted :( This is the best I an do.
> 
> I am able to fix the issue by installing `ansible-pylibssh` using `pip3 install ansible-pylibssh`. After removing it, I again get the error.
> 
> The fallback when `pylibssh` isn't available is `paramiko`, which I've been using since I started using this module.
> 
> Interestingly, `ansible-pylibssh` was removed from the `requirements.txt` of this module in this commit: [54e0d62](https://github.com/vyos/vyos.vyos/commit/54e0d62b65deb4993f73571b7ea0ca9138f79e03)
> 
> My guess is that your environment is configured to use `ansible-pylibssh`.
> 
> To verify, please force Ansible to use paramiko:
> 
> Add `ansible_network_cli_ssh_type=paramiko` to your `hosts.ini` file, in the vars section, and run the tests again. I expect this to fail, just like it did for me.
> 
> > If the tests still succeed, Try with `ansible_network_cli_ssh_type=test` to verify if the variable is actually used. This should fail in just 5 seconds, when ansible launches the network cli script

Thanks mate - nice spotting.
I will try with the paramiko, however, I believe, if it fails, a combination of paramiko and python libs must have been mutating the stdout, thus causing the string comparison to fail.
I am using stdout in this integration testing

RubenNL commented:
PR created: https://github.com/vyos/vyos.vyos/pull/409
